File: leetcode/1046.py
from bisect import bisect_left

class Solution:
	def lastStoneWeight(self, stones: List[int]) -> int:
		stones.sort()
		while len(stones) > 1:
			a = stones.pop()
			b = stones.pop()
			if a > b:
				stones.insert(bisect_left(stones, a - b), a - b)
		if len(stones) == 0:
			return 0
		return stones[0]

File: leetcode/1046.rs
// LeetCode/1046 - Last Stone Weight

use std::collections::BinaryHeap;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut heap = BinaryHeap::from(stones);
        while heap.len() > 1 {
            let y = heap.pop().unwrap();
            let x = heap.pop().unwrap();

            if x == y {
                continue;
            }

            let smashed = y - x;
            heap.push(smashed);
        }
        return if heap.len() == 0 {
            0
        } else {
            heap.pop().unwrap()
        };
    }
}
